src/key-value-engine/rocksdb.md

```diff
@@ -0,0 +1,155 @@
+lifetime) of each row. If we have no idea whether a region contains
+some garbage versions or not, we have to check all regions
+periodically. To skip unnecessary garbage collection, we record some
+MVCC statistics (e.g. the number of rows and the number of versions)
+in each SST file. So before checking every region row by row, we check
+the table properties to see if it is necessary to do garbage
+collection on a region.
+
+## CompactRange
+
+From time to time, some regions may contain a lot of tombstone entries
+because of GC or other delete operations. Tombstone entries are not
+good for scan performance and waste disk space as well.
+
+So with the `TableProperties` feature, we can check every region
+periodically to see if it contains a lot of tombstones. If it does, we
+will compact the region range manually to drop tombstone entries and
+release disk space.
+
+We also use `CompactRange` to recover RocksDB from some mistakes like
+incompatible table properties across different TiKV versions.
+
+## [EventListener](https://github.com/facebook/rocksdb/wiki/EventListener)
+
+`EventListener` allows us to listen to some special events, like
+flush, compaction or write stall condition change. When a specific
+event is triggered or finished, RocksDB will invoke our callbacks with
+some information about the event.
+
+TiKV listens to the compaction event to observe the region size
+changes. As mentioned above, we calculate the approximate size of each
+region from the table properties. The approximate size will be
+recorded in the memory so that we don't need to calculate it again and
+again if nothing has changed. However, during compactions, some
+entries are dropped so the approximate size of some regions should be
+updated. That's why we listen to the compaction events and recalculate
+the approximate size of some regions when necessary.
+
+## [IngestExternalFile](https://github.com/facebook/rocksdb/wiki/Creating-and-Ingesting-SST-files)
+
+RocksDB allows us to generate an SST file outside and then ingest the
+file into RocksDB directly. This feature can potentially save a lot of
+IO because RocksDB is smart enough to ingest a file to the bottom
+level if possible, which can reduce write amplification because the
+ingested file doesn't need to be compacted again and again.
+
+We use this feature to handle Raft snapshot. For example, when we want
+to add a replica to a new server. We can first generate a snapshot
+file from another server and then send the file to the new
+server. Then the new server can ingest that file into its RocksDB
+directly, which saves lots of work!
+
+We also use this feature to import a huge amount of data into TiKV. We
+have some tools to generate sorted SST files from different data
+sources and then ingest those files into different TiKV servers. This
+is super fast compared to writing key-values to a TiKV cluster in the
+usual way.
+
+## DeleteFilesInRange
```

Comment:
```suggestion
## DeleteFilesInRanges
```

Reply:
We actually use both, and I think `DeleteFilesInRange` is more general and easy to understand.